File: sdks/java/src/main/java/com/syros/SyrosClient.java
package com.syros;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;
import okhttp3.*;
import okhttp3.ws.WebSocket;
import okhttp3.ws.WebSocketListener;
import okio.ByteString;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Syros Java SDK
 * 
 * Provides a Java interface to interact with Syros,
 * including REST APIs and WebSocket communication.
 */
public class SyrosClient {
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final String restUrl;
    private final String wsUrl;
    private WebSocket webSocket;

    public SyrosClient() {
        this("http://localhost:8080", "ws://localhost:8080/ws");
    }

    public SyrosClient(String restUrl, String wsUrl) {
        this.restUrl = restUrl;
        this.wsUrl = wsUrl;
        this.objectMapper = new ObjectMapper();
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    // --- REST API Methods ---

    /**
     * Health check endpoint
     */
    public JsonNode healthCheck() throws IOException {
        return sendRestRequest("GET", "/health", null);
    }

    /**
     * Acquire a distributed lock
     */
    public JsonNode acquireLock(String key, String owner, Long ttlSeconds, String metadata) throws IOException {
        Map<String, Object> payload = new HashMap<>();
        payload.put("key", key);
        payload.put("owner", owner);
        if (ttlSeconds != null) payload.put("ttl_seconds", ttlSeconds);
        if (metadata != null) payload.put("metadata", metadata);

        return sendRestRequest("POST", "/api/v1/locks", payload);
    }

    /**
     * Release a distributed lock
     */
    public JsonNode releaseLock(String key, String lockId, String owner) throws IOException {
        Map<String, Object> payload = new HashMap<>();
        payload.put("lock_id", lockId);
        payload.put("owner", owner);

        return sendRestRequest("DELETE", "/api/v1/locks/" + key, payload);
    }

    /**
     * Get lock status
     */
    public JsonNode getLockStatus(String key) throws IOException {
        return sendRestRequest("GET", "/api/v1/locks/" + key + "/status", null);
    }

    /**
     * Start a saga
     */
    public JsonNode startSaga(String name, Object[] steps, Map<String, String> metadata) throws IOException {
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("steps", steps);
        if (metadata != null) payload.put("metadata", metadata);

        return sendRestRequest("POST", "/api/v1/sagas", payload);
    }

    /**
     * Get saga status
     */
    public JsonNode getSagaStatus(String sagaId) throws IOException {
        return sendRestRequest("GET", "/api/v1/sagas/" + sagaId + "/status", null);
    }

    /**
     * Append event to event store
     */
    public JsonNode appendEvent(String streamId, String eventType, Object data, Map<String, String> metadata) throws IOException {
        Map<String, Object> payload = new HashMap<>();
        payload.put("stream_id", streamId);
        payload.put("event_type", eventType);
        payload.put("data", data);
        if (metadata != null) payload.put("metadata", metadata);

        return sendRestRequest("POST", "/api/v1/events", payload);
    }

    /**
     * Get events from event store
     */
    public JsonNode getEvents(String streamId) throws IOException {
        return sendRestRequest("GET", "/api/v1/events/" + streamId, null);
    }

    /**
     * Set cache value
     */
    public JsonNode setCache(String key, Object value, Long ttlSeconds, String[] tags) throws IOException {
        Map<String, Object> payload = new HashMap<>();
        payload.put("value", value);
        if (ttlSeconds != null) payload.put("ttl_seconds", ttlSeconds);
        if (tags != null) payload.put("tags", tags);

        return sendRestRequest("POST", "/api/v1/cache/" + key, payload);
    }

    /**
     * Get cache value
     */
    public JsonNode getCache(String key) throws IOException {
        return sendRestRequest("GET", "/api/v1/cache/" + key, null);
    }

    /**
     * Delete cache value
     */
    public JsonNode deleteCache(String key) throws IOException {
        return sendRestRequest("DELETE", "/api/v1/cache/" + key, null);
    }

    /**
     * Get metrics
     */
    public JsonNode getMetrics() throws IOException {
        return sendRestRequest("GET", "/metrics", null);
    }

    // --- WebSocket API Methods ---

    /**
     * Connect to WebSocket
     */
    public void connectWebSocket(WebSocketListener listener) {
        Request request = new Request.Builder()
                .url(wsUrl)
                .build();

        webSocket = httpClient.newWebSocket(request, listener);
    }

    /**
     * Send WebSocket message
     */
    public void sendWebSocketMessage(String type, Object data) throws IOException {
        if (webSocket != null) {
            Map<String, Object> message = new HashMap<>();
            message.put("type", type);
            message.put("data", data);
            
            String jsonMessage = objectMapper.writeValueAsString(message);
            webSocket.send(ByteString.encodeUtf8(jsonMessage));
        }
    }

    /**
     * Disconnect WebSocket
     */
    public void disconnectWebSocket() throws IOException {
        if (webSocket != null) {
            webSocket.close(1000, "Client disconnect");
            webSocket = null;
        }
    }

    // --- Private Helper Methods ---

    private JsonNode sendRestRequest(String method, String path, Object data) throws IOException {
        String url = restUrl + path;
        Request.Builder requestBuilder = new Request.Builder().url(url);

        if (data != null) {
            String jsonData = objectMapper.writeValueAsString(data);
            RequestBody body = RequestBody.create(jsonData, MediaType.get("application/json; charset=utf-8"));
            
            if ("POST".equals(method)) {
                requestBuilder.post(body);
            } else if ("DELETE".equals(method)) {
                requestBuilder.delete(body);
            } else if ("PUT".equals(method)) {
                requestBuilder.put(body);
            }
        } else {
            if ("DELETE".equals(method)) {
                requestBuilder.delete();
            }
        }

        Request request = requestBuilder.build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("HTTP " + response.code() + ": " + response.message());
            }
            
            String responseBody = response.body().string();
            return objectMapper.readTree(responseBody);
        }
    }

    /**
     * Close the client and release resources
     */
    public void close() throws IOException {
        disconnectWebSocket();
        httpClient.dispatcher().executorService().shutdown();
    }
}

File: sdks/go/syros_client.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"github.com/gorilla/websocket"
)

// SyrosClient provides a Go interface to interact with Syros
type SyrosClient struct {
	httpClient *http.Client
	restURL    string
	wsURL      string
	wsConn     *websocket.Conn
}

// NewSyrosClient creates a new Syros client with default settings
func NewSyrosClient() *SyrosClient {
	return NewSyrosClientWithURLs("http://localhost:8080", "ws://localhost:8080/ws")
}

// NewSyrosClientWithURLs creates a new Syros client with custom URLs
func NewSyrosClientWithURLs(restURL, wsURL string) *SyrosClient {
	return &SyrosClient{
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		restURL: restURL,
		wsURL:   wsURL,
	}
}

// HealthCheck performs a health check
func (c *SyrosClient) HealthCheck() (map[string]interface{}, error) {
	return c.sendRestRequest("GET", "/health", nil)
}

// AcquireLock acquires a distributed lock
func (c *SyrosClient) AcquireLock(key, owner string, ttlSeconds *int64, metadata *string) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"key":   key,
		"owner": owner,
	}
	if ttlSeconds != nil {
		payload["ttl_seconds"] = *ttlSeconds
	}
	if metadata != nil {
		payload["metadata"] = *metadata
	}

	return c.sendRestRequest("POST", "/api/v1/locks", payload)
}

// ReleaseLock releases a distributed lock
func (c *SyrosClient) ReleaseLock(key, lockID, owner string) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"lock_id": lockID,
		"owner":   owner,
	}

	return c.sendRestRequest("DELETE", "/api/v1/locks/"+key, payload)
}

// GetLockStatus gets the status of a lock
func (c *SyrosClient) GetLockStatus(key string) (map[string]interface{}, error) {
	return c.sendRestRequest("GET", "/api/v1/locks/"+key+"/status", nil)
}

// StartSaga starts a new saga
func (c *SyrosClient) StartSaga(name string, steps []map[string]interface{}, metadata map[string]string) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"name":  name,
		"steps": steps,
	}
	if metadata != nil {
		payload["metadata"] = metadata
	}

	return c.sendRestRequest("POST", "/api/v1/sagas", payload)
}

// GetSagaStatus gets the status of a saga
func (c *SyrosClient) GetSagaStatus(sagaID string) (map[string]interface{}, error) {
	return c.sendRestRequest("GET", "/api/v1/sagas/"+sagaID+"/status", nil)
}

// AppendEvent appends an event to the event store
func (c *SyrosClient) AppendEvent(streamID, eventType string, data interface{}, metadata map[string]string) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"stream_id":  streamID,
		"event_type": eventType,
		"data":       data,
	}
	if metadata != nil {
		payload["metadata"] = metadata
	}

	return c.sendRestRequest("POST", "/api/v1/events", payload)
}

// GetEvents gets events from the event store
func (c *SyrosClient) GetEvents(streamID string) (map[string]interface{}, error) {
	return c.sendRestRequest("GET", "/api/v1/events/"+streamID, nil)
}

// SetCache sets a value in the cache
func (c *SyrosClient) SetCache(key string, value interface{}, ttlSeconds *int64, tags []string) (map[string]interface{}, error) {
	payload := map[string]interface{}{
		"value": value,
	}
	if ttlSeconds != nil {
		payload["ttl_seconds"] = *ttlSeconds
	}
	if tags != nil {
		payload["tags"] = tags
	}

	return c.sendRestRequest("POST", "/api/v1/cache/"+key, payload)
}

// GetCache gets a value from the cache
func (c *SyrosClient) GetCache(key string) (map[string]interface{}, error) {
	return c.sendRestRequest("GET", "/api/v1/cache/"+key, nil)
}

// DeleteCache deletes a value from the cache
func (c *SyrosClient) DeleteCache(key string) (map[string]interface{}, error) {
	return c.sendRestRequest("DELETE", "/api/v1/cache/"+key, nil)
}

// GetMetrics gets Prometheus metrics
func (c *SyrosClient) GetMetrics() (string, error) {
	resp, err := c.sendRestRequestRaw("GET", "/metrics", nil)
	if err != nil {
		return "", err
	}
	return resp, nil
}

// ConnectWebSocket connects to the WebSocket
func (c *SyrosClient) ConnectWebSocket() error {
	u, err := url.Parse(c.wsURL)
	if err != nil {
		return err
	}

	c.wsConn, _, err = websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		return err
	}

	fmt.Println("WebSocket connected to", c.wsURL)
	return nil
}

// SendWebSocketMessage sends a message through WebSocket
func (c *SyrosClient) SendWebSocketMessage(messageType string, data interface{}) error {
	if c.wsConn == nil {
		return fmt.Errorf("WebSocket not connected")
	}

	message := map[string]interface{}{
		"type": messageType,
		"data": data,
	}

	return c.wsConn.WriteJSON(message)
}

// ListenWebSocket listens for WebSocket messages
func (c *SyrosClient) ListenWebSocket(onMessage func(string)) error {
	if c.wsConn == nil {
		return fmt.Errorf("WebSocket not connected")
	}

	for {
		_, message, err := c.wsConn.ReadMessage()
		if err != nil {
			return err
		}
		onMessage(string(message))
	}
}

// DisconnectWebSocket disconnects from WebSocket
func (c *SyrosClient) DisconnectWebSocket() error {
	if c.wsConn != nil {
		err := c.wsConn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
		if err != nil {
			return err
		}
		c.wsConn.Close()
		c.wsConn = nil
		fmt.Println("WebSocket disconnected")
	}
	return nil
}

// Close closes the client and releases resources
func (c *SyrosClient) Close() error {
	c.DisconnectWebSocket()
	return nil
}

// Private helper methods

func (c *SyrosClient) sendRestRequest(method, path string, data interface{}) (map[string]interface{}, error) {
	resp, err := c.sendRestRequestRaw(method, path, data)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal([]byte(resp), &result); err != nil {
		return nil, err
	}

	return result, nil
}

func (c *SyrosClient) sendRestRequestRaw(method, path string, data interface{}) (string, error) {
	url := c.restURL + path
	var body io.Reader

	if data != nil {
		jsonData, err := json.Marshal(data)
		if err != nil {
			return "", err
		}
		body = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequestWithContext(context.Background(), method, url, body)
	if err != nil {
		return "", err
	}

	if data != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	return string(responseBody), nil
}
